### ISSUE TITLE
Synapse is not limiting the number of EDUs in a transaction

### ISSUE THREAD

**Issue by turt2live** [matrix-org/synapse#3951]
It looks like https://github.com/matrix-org/synapse/pull/3805 tries to, however there's some evidence that the EDU limiting doesn't work:

```
homeserver_1 - 2018-09-25 14:48:16,126 - synapse.federation.transport.server - 336 - INFO - PUT-381748 - Received txn 1537790360731 from matrix.org. (PDUs: 50, EDUs: 166 - DROPPED)
```

(EDUs are 'dropped' because of a change on t2bot.io - this doesn't change the number of EDUs received from matrix.org)

**Comment by richvdh** [comment]
yeah, there are other EDUs which aren't included in the clip in in #3805. I'll handle this as part of my work on #4671.

**Comment by anoadragon453** [comment]
@richvdh Was this handled?

**Comment by richvdh** [comment]
Not really, I'm afraid. The original solution to #4730 got thrown away, and it's still possible for synapse to spam out more than 100 EDUs in a transaction.

**Comment by turt2live** [comment]
This is fairly urgent imo now that https://github.com/matrix-org/synapse/pull/4513 is a thing - servers cannot federate with each other because the transactions are being turned away.

**Comment by superboum** [comment]
I am currently encountering the problem described by @turt2live :

```
2019-05-04 06:16:57,274 - synapse.federation.sender.transaction_manager - 87 - INFO - federation_transaction_transmission_loop-678370- TX [matrix.livda.fr] {1556878459543} Sending transaction [1556878459543], (PDUs: 0, EDUs: 103)
2019-05-04 06:16:57,338 - synapse.http.matrixfederationclient - 356 - INFO - federation_transaction_transmission_loop-678370- {PUT-O-448448} [matrix.livda.fr] Sending request: PUT matrix://matrix.livda.fr/_matrix/federation/v1/send/1556878459543; timeout 60.000000s
2019-05-04 06:16:57,339 - synapse.http.federation.matrix_federation_agent - 373 - INFO - federation_transaction_transmission_loop-678370- Connecting to matrix.livda.fr:443
[...]
2019-05-04 06:17:08,409 - synapse.http.matrixfederationclient - 388 - INFO - federation_transaction_transmission_loop-678370- {PUT-O-448448} [matrix.livda.fr] Got response headers: 400 Bad Request
2019-05-04 06:17:08,410 - synapse.http.matrixfederationclient - 472 - WARNING - federation_transaction_transmission_loop-678370- {PUT-O-448448} [matrix.livda.fr] Request failed: PUT matrix://matrix.livda.fr/_matrix/federation/v1/send/1556878459543: HttpResponseException("400: b'Bad Request'",)
2019-05-04 06:17:08,411 - synapse.federation.sender.transaction_manager - 123 - INFO - federation_transaction_transmission_loop-678370- TX [matrix.livda.fr] {1556878459543} got 400 response
```

The problem doesn't seem to disappear by itself.

**Comment by richvdh** [comment]
Fixed by #5138
